src/coreclr/vm/riscv64/asmhelpers.S

```diff
@@ -539,6 +539,68 @@ LOCAL_LABEL(JIT_GetSharedGCStaticBase_SingleAppDomain_CallHelper):
     call  JIT_GetSharedGCStaticBase_Helper
 LEAF_END JIT_GetSharedGCStaticBase_SingleAppDomain, _TEXT
 
+// Make sure the `FaultingExceptionFrame_StackAlloc` is 16-byte aligned.
+#define FaultingExceptionFrame_StackAlloc (SIZEOF__GSCookie + SIZEOF__FaultingExceptionFrame + 0x8)
+#define FaultingExceptionFrame_FrameOffset SIZEOF__GSCookie
+
+.macro GenerateRedirectedStubWithFrame stub, target
+
+    //
+    // This is the primary function to which execution will be redirected to.
+    //
+    NESTED_ENTRY \stub, _TEXT, NoHandler
+
+        //
+        // IN: ra: original IP before redirect
+        //
+
+        PROLOG_SAVE_REG_PAIR_INDEXED  fp, ra, 16
+
+        // alloc stack for FaultingExceptionFrame.
+        addi  sp, sp, -FaultingExceptionFrame_StackAlloc
+
+        // stack must be 16 bytes aligned
+        CHECK_STACK_ALIGNMENT
```

Comment:
I think it's not correct. Original ARM64 code always align stack, but `CHECK_STACK_ALIGNMENT` works only on debug mode.

Reply:
The line:542:543,
```
// Make sure the `FaultingExceptionFrame_StackAlloc` is 16-byte aligned.
#define FaultingExceptionFrame_StackAlloc (SIZEOF__GSCookie + SIZEOF__FaultingExceptionFrame + 0x8)
```
we define the `FaultingExceptionFrame_StackAlloc`  is 16-byte aligned while the ARM64 is 8-byte aligned.

Reply:
Yes, as @shushanhf said, this is a small optimization.

Reply:
May be I don't understand your idea clearly, but comment in ARM64 says:
```
; At this point, the stack maybe misaligned if the thread abort was asynchronously
; triggered in the prolog or epilog of the managed method. For such a case, we must
; align the stack before calling into the VM.
;
; Runtime check for 16-byte alignment.
```
That means that stack might be already misaligned before `addi sp, sp, -FaultingExceptionFrame_StackAlloc`.
@jkotas am I right?

Reply:
Yes, you are right. The thread abort stops the thread, checks whether the thread is inside managed code. If the thread is inside managed code, it patches the thread IP to point to this method. If the thread happened to be stopped in the middle of epilog or prolog, the stack is going to be misaligned.

Reply:
First, for LoongArch64, we always keep the sp is 16-byte aligned, even in the manged code.
Second, during the processes of the abort by vm, I think the vm can get the sp whether is 16-byte aligned by the the FaultingExceptionFrame and should throw the error for unaligned sp.

Reply:
> May be I don't understand your idea clearly, but comment in ARM64 says:
> 
> ```
> ; At this point, the stack maybe misaligned if the thread abort was asynchronously
> ; triggered in the prolog or epilog of the managed method. For such a case, we must
> ; align the stack before calling into the VM.
> ;
> ; Runtime check for 16-byte alignment.
> ```

Where is the sp is not 16 byte aligned within the prolog or epilog ?
For LoongArch64, even the stack is too large while the sp should be adjust by two or more times, every adjusting the sp is 16 byte aligned. I think the ARM64 should be also this.

I think the prolog and epilog for ARM64, RISCV64 and LoongArch64, they should keep the sp always 16-byte aligned during the adjusting the sp.

Reply:
Ok, I can see that the sp should be always aligned given the codegen used for prologs/epilogs.

Reply:
> For LoongArch64, even the stack is too large while the sp should be adjusted by two or more times, every time adjusting the sp is 16 byte aligned. I think the ARM64 should be also this.

It doesn't work for ARM64, e.g. there is instruction like `str x1, [sp], 8` which change stack to `8`. Situation with RISC-V and LA64 is better since we change `sp` manualy, but can we grantee that it will work that way now and with future changes?

Reply:
> > For LoongArch64, even the stack is too large while the sp should be adjusted by two or more times, every time adjusting the sp is 16 byte aligned. I think the ARM64 should be also this.
> 
> It doesn't work for ARM64, e.g. there is instruction like `str x1, [sp], 8` which change stack to `8`. Situation with RISC-V and LA64 is better since we change `sp` manualy, but can we grantee that it will work that way now and with future changes?

This instruction is not used within the manged code and I think we should not use this instruction to change the sp.
The increases should be only used when store paried and keep 16-byte aligned when changing the sp.

The proper way is first allock the enough stack space and then store, rather than store one by one while changing the sp 8-byte each time.

This instruction should only be used by pointer except the sp pointer.

Now the prolog and epilog don't change the sp by this way.

Reply:
According to RISC-V psABI, it required `$sp` to be 16-byte aligned. I think if the `$sp` here is not 16-byte aligned, then the `$sp` is likely to be incorrect already.

Reply:
> It doesn't work for ARM64, e.g. there is instruction like `str x1, [sp], 8` which change stack to `8`. Situation with RISC-V and LA64 is better since we change `sp` manualy, but can we grantee that it will work that way now and with future changes?

I didn't modify the code for ARM64 in this commit, maybe we just need to focus on how to make LoongArch64 and RISC-V work properly.